android/greengrass-nucleus-android/src/main/java/com/aws/greengrass/android/managers/AndroidBaseComponentControl.java

```diff
@@ -0,0 +1,707 @@
+         *
+         * @return exit code of component
+         * @throws InterruptedException when current thread was interrupted
+         */
+        private int waitCompletion() throws InterruptedException {
+            thread.join();
+            return exitCode.get();
+        }
+
+        /**
+         * Send STOP command if communication channel is opened, quit looper and join with it thread.
+         *
+         * @throws InterruptedException when thread has been interrupted
+         */
+        private void terminate() throws InterruptedException {
+            logger.atDebug().kv(PACKAGE_NAME, packageName).kv(CLASS_NAME, className)
+                    .log("Request terminate component");
+            requestUnbind();
+            thread.join();
+        }
+
+        private void terminateComponent(long msTimeout) throws RuntimeException, InterruptedException {
+            sendExitRequest(msTimeout);
+            terminate();
+        }
+
+        /**
+         * Send EXIT command and waiting for response for time out.
+         *
+         */
+        private void sendExitRequest(long msTimeout) throws RuntimeException {
+            Messenger messenger = messengerService.get();
+            Messenger replyTo = replyMessenger.get();
+            if (messenger != null && replyTo != null) {
+                try {
+                    // Prepare CompletableFuture to wait for the component to exit
+                    CompletableFuture componentTerminatedFuture = new CompletableFuture();
+                    componentTerminatedFutureRef.set(componentTerminatedFuture);
+
+                    // Send exit request
+                    Message msg = Message.obtain(null, LIFECYCLE_MSG_REQUEST_EXIT);
+                    msg.replyTo = replyTo;
+                    messenger.send(msg);
+
+                    // Wait for component to exit
+                    if (msTimeout > 0) {
+                        // Wait for the specified amount of time
+                        componentTerminatedFuture.get(msTimeout, TimeUnit.MILLISECONDS);
+                        logger.atDebug().kv(PACKAGE_NAME, packageName).kv(CLASS_NAME, className)
+                                .log("Component terminated within timeout");
+                    } else {
+                        // Wait indefinitely
+                        componentTerminatedFuture.get();
+                        logger.atDebug().kv(PACKAGE_NAME, packageName).kv(CLASS_NAME, className)
+                                .log("Component terminated normally");
+                    }
+                } catch (RemoteException | IllegalStateException e) {
+                    logger.atError().kv(PACKAGE_NAME, packageName).kv(CLASS_NAME, className)
+                            .setCause(e).log("Unable to terminate the service");
+                    throw new RuntimeException("Unable to terminate the service", e);
+                } catch (ExecutionException | InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)